Store Orders: Update order list filters & add searching

ryelle opened the pull request:
Fixes #16288, Fixes #16443 

This PR adds real order status filters to the nav bar above the orders table. Currently this PR only shows "processing" and "completed"– eventually this will change to "All orders", "Awaiting fulfillment", and "Awaiting payment" (but we need https://github.com/woocommerce/wc-api-dev/pull/31 finished first), so this PR is more about laying out the filter functionality.

This also adds searching to the filter bar.

![screen shot 2017-07-26 at 1 02 24 pm](https://user-images.githubusercontent.com/541093/28633697-afbef0de-7202-11e7-9eee-9b6fc8cf154e.png)
![screen shot 2017-07-26 at 1 02 41 pm](https://user-images.githubusercontent.com/541093/28633707-b92e552e-7202-11e7-8769-dd300580ea7a.png)

It adds a placeholder while items are loading:

<img width="786" alt="screen shot 2017-07-25 at 12 18 07 pm" src="https://user-images.githubusercontent.com/541093/28582433-5b4791d4-7133-11e7-9a8f-92d8c0a75130.png">

And empty states when there are no orders for a status or search:

![screen shot 2017-07-26 at 1 03 42 pm](https://user-images.githubusercontent.com/541093/28633764-e0719d58-7202-11e7-8f6b-fe660225687a.png)

It will alternately say "You don't have any orders awaiting payment." or "You don\'t have any orders awaiting fulfillment." for those statuses (once they're properly hooked up).

Part of adding searching was also tracking search terms in the UI redux state, and it turns out we don't need to track status in redux state because it's tracked by the path itself. So there's also a change to swap the status-related code to search term tracking.

**To test**

- Visit `/store/orders/:site`
- Click around on the filters, make sure correct orders are showing
- If you have more than 50 orders, check that pagination + filters work (or tweak the per_page number in `<Pagination…` in `order-list.js` and in `state/sites/orders/utils.js`)
  - Switching the status should jump you back to page 1 of that status
  - but changing page should not change the status
- Search through your orders, pagination should still work
- Run `npm test` to make sure that the status/search switch works.

a8ci18n commented on client/extensions/woocommerce/index.js:
```diff
@@ -64,6 +64,12 @@ const getStorePages = () => {
 		{
+			container: Orders,
+			configKey: 'woocommerce/extension-orders',
+			documentTitle: translate( 'Orders' ),
```
Hi! I've found a possible matching string that has already been translated **100** times:
`translate( 'Order' )` *ES Score: 6*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/orders-status-filtering)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/extensions/woocommerce/app/orders/orders-filter-nav.js:
```diff
@@ -0,0 +1,42 @@
+						{ translate( 'Processing' ) }
+					</NavItem>
+					<NavItem path={ getLink( '/store/orders/completed/:site', site ) } selected={ 'completed' === status }>
+						{ translate( 'Completed' ) }
```
Hi! I've found a possible matching string that has already been translated **52** times:
`translate( '&laquo; completed' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/orders-status-filtering)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

allendav commented on client/extensions/woocommerce/app/orders/orders-list.js:
```diff
@@ -99,7 +104,20 @@ class Orders extends Component {
 
-	renderOrderItems = ( order, i ) => {
+	renderPlacholder = () => {
+		return range( 5 ).map( ( i ) => {
```
Oooh - nice way to do placeholders

allendav replied:
But let's call it `renderPlaceholder`

allendav commented on client/extensions/woocommerce/app/orders/orders-list.js:
```diff
@@ -183,11 +203,11 @@ class Orders extends Component {
 
 export default connect(
-	state => {
+	( state, props ) => {
```
Optional suggestion - pull this out into a mapstatetoprops method like we do elsewhere in the extension

allendav replied:
Similar comment for dispatch below - totally optional

ryelle commented:
> For the placeholder, is there any way we could count the orders and display an equal number of rows?

Unfortunately no, as soon as we know how many orders there are we have the orders and can display them :)

I had been working on some other fixes and adding search in another PR, but it's getting too complicated to keep them separate 😣 — I'm putting this one on "in progress" and merging the other PR in, I'll update and re-ping when this should be re-reviewed.

ryelle commented on client/extensions/woocommerce/app/orders/orders-list.js:
```diff
@@ -46,16 +47,28 @@ class Orders extends Component {
+			query.page = 1;
+			query.search = '';
+		}
+		this.props.fetchOrders( newProps.siteId, query );
```
I think these lines ^ need a re-think, since I need to reset some parts of the query based on what just changed, but I'm not sure what to do - any ideas would be welcome :)

ryelle commented:
I've updated the description with more testing instructions & screenshots, now that this PR also includes searching, and updated empty-content (no orders found) states.

jameskoster commented:
> Unfortunately no, as soon as we know how many orders there are we have the orders and can display them

Thought that might be the case, nvm then :)

I just thought, could we make the placeholder consistent with [products](https://cldup.com/uqbae31MXM.gif)? That gif doesn't display it very well but there's a full-height fading div. It's not perfect but the inconsistency in rows bugs me.

This is not a big deal though, we can definitely tweak later.

ryelle commented:
@jameskoster sure, I'm happy to do placeholder tweaks in another PR. Is design ✅ otherwise?

jameskoster commented:
Couple of small things that you might want to do in the placeholder PR;

The EmptyContent message for the preset filters could be more contextual;

Also, ideally those EmptyContent's would appear alone, outside the table, with no thead.

![empty-orders](https://cldup.com/0tJ2iYv3tt-2000x2000.png)

justinshreve commented on client/extensions/woocommerce/state/ui/orders/test/reducer.js:
```diff
@@ -36,49 +36,62 @@ describe( 'reducer', () => {
 	} );
 
-	it( 'should should update the current status when changed', () => {
+	it( 'should should update the current search when changed', () => {
```
Double `should`

justinshreve commented on client/extensions/woocommerce/app/orders/orders-filter-nav.js:
```diff
@@ -0,0 +1,66 @@
+					onSearchClose={ this.clearSearch }
+					placeholder={ translate( 'Search orders' ) }
+					analyticsGroup="Orders"
+					delaySearch={ true }
```
This can just be shortened to `delaySearch`.

ryelle commented:
@kellychoffman Which pagination style is the one we should be using? Products adds some custom styles to look like ^, orders uses the default pagination which looks like:

<img width="738" alt="screen shot 2017-07-31 at 2 01 17 pm" src="https://user-images.githubusercontent.com/541093/28790946-cb14f744-75f8-11e7-86e9-44c209ab9822.png">

The only other place in calypso this is being used (right now) is on the comment followers page, where it's inside a card, and it looks like this:

<img width="851" alt="screen shot 2017-07-31 at 2 02 48 pm" src="https://user-images.githubusercontent.com/541093/28791024-0bf688d6-75f9-11e7-9495-3ecbcd040723.png">

kellychoffman commented:
> Which pagination style is the one we should be using? Products adds some custom styles

Let's use the styles products is using if that's okay. In meetup mode right now, but let me know if that means I need to refactor styles.

ryelle commented:
Since I'm not technically adding pagination in this PR, I think those style updates should take place in a follow-up - @justinshreve, is this OK to merge, or do you have any other feedback?

justinshreve commented:
@ryelle I don't have any further feedback. You can handle styles in a separate PR.

kellychoffman commented:
Awesome!
